go/tasks/pluginmachinery/utils/dns.go

```diff
@@ -0,0 +1,36 @@
+package utils
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/encoding"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+var dns1123InvalidRegex = regexp.MustCompile("[^-a-z0-9]")
+var camelCaseRegex = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+const maxUniqueIDLength = 20
+
+// ConvertToDNS1123CompatibleString converts a string that doesn't conform to the definition of a label in DNS (RFC 1123) to a string that conforms.
+func ConvertToDNS1123CompatibleString(name string) string {
```

Comment:
Is this strictly backward compatible? I'm worried about upgrade scenarios, if I've workflows in flight (pun intended) that launch a pod with name X, things will start failing if we start looking for pods under a different name X'...

Reply:
Yes it will break things if the pod name changes mid-flight. The assumption here is that if a name violates 1123, it will never start properly to begin with (and if it conforms to 1123, nothing changes) ... though, it's still a scary thought that it might break something. Would you recommend some test scenarios that I can run through?

Reply:
One potential change is to run the validation here to make sure that if it would have been accepted, we don't try to touch it... can you triple check that DNS1123 is what they validate pod names against?

```suggestion
func ConvertToDNS1123CompatibleString(name string) string {

    if validation.IsDNS1123(name) {
        return name
    }
```

Reply:
No official literature on this subject matter. https://kubernetes.io/docs/concepts/overview/working-with-objects/names/ vaguely mentioned the four types of commonly used name constraints for resources: "DNS Subdomain Names", "RFC 1123 Label Names", "RFC 1035 Label Names", and "Path Segment Names". The closest I could find (Google) was this page from Oracle https://docs.oracle.com/en/operating-systems/olcne/1.1/orchestration/kubectl-run.html that claims `Deployment, pod and service names conform to a requirement to match a DNS-1123 label. These must consist of lower case alphanumeric characters or -, and must start and end with an alphanumeric character. The regular expression that is used to validate names is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'. If you use a name for your deployment that does not validate, an error is returned.`. This statement is either stale or inaccurate imo. In the current implementation of K8s, the pod name is validated against "DNS Subdomain Names" as shown here https://github.com/kubernetes/kubernetes/blob/4451138bfa784942e59dc092fc8d707e73eaa284/pkg/apis/core/validation/validation.go#L3969 and here https://github.com/kubernetes/kubernetes/blob/4451138bfa784942e59dc092fc8d707e73eaa284/pkg/apis/core/validation/validation.go#L229. The error message "a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character" matches the error message in https://github.com/flyteorg/flyte/issues/1172 . 
To conclude: no written statement on pod name validation, but evidences point to "DNS Subdomain Names" which is 253 characters max.